### PR TITLE
perf(ingester): catalog bloom filter for partition IDs

### PR DESCRIPTION
**Pull request by domodwyer**
This practically eliminates the catalog queries for partitions created since #7963 was merged.

This uses about 2MiB of memory when saturated, and decreasing as the partitions age out.

Ideally we'd be able to remove this entirely in the future.

Part of https://github.com/influxdata/influxdb_iox/issues/8066

---

* perf: minimise partition catalog queries (95ac393d5)
      
      This commit implements a PartitionProvider decorator that
      probabilistically determines if a partition is going to be a an
      "old-style" row-addressed partition created prior to #7963, or a
      "new-style" hash-addressed partition created after using a fast,
      space-efficient, compressed bloom filter.
      
      If a partition is identified as a new-style, hash-addressed partition,
      the PartitionData is immediately initialised using the deterministic
      hash ID without performing a catalog query at all.
      
      If a partition is identified as an old-style, row-addressed partition, a
      catalog query is performed to resolve the row ID as it would without
      this filter.
      
      A new-style, hash-addressed partition may sometimes be incorrectly
      identified as a row-addressed partition, causing a spurious catalog
      query, which is then correctly identified as a hash-addressed partition.
      This is tuned to happen ~1-0.1% of the time, eliminating 99% to 99.9% of
      unnecessary catalog queries.

* chore: regenerate workspace-hack (011200af3)

### CODE REVIEW

File: workspace-hack/Cargo.toml [hunk 1]
```diff
@@ -61,7 +61,7 @@ parquet = { version = "43", features = ["experimental", "object_store"] }
 petgraph = { version = "0.6" }
 phf_shared = { version = "0.11" }
 predicates = { version = "3" }
-proptest = { version = "1", default-features = false, features = ["std"] }
+proptest = { version = "1" }
```
Comment (crepererum) [inline workspace-hack/Cargo.toml:64]: could you please NOT include the default features? IIRC the fork system that proptest comes with has been led to deadlocks in the past

Comment (domodwyer) [reply workspace-hack/Cargo.toml:64]: Ah sure - is there an issue for the deadlock?

File: ingester/src/buffer_tree/partition/resolver/old_filter.rs [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+            partition_key in arbitrary_partition_key(),
+            row_id in arbitrary_row_id(),
+            has_hash_id in any::<bool>(),
+        ) -> Partition {
+            let hash_id = match has_hash_id {
+                true => Some(PartitionHashId::new(table_id, &partition_key)),
+                false => None,
+            };
+
+            Partition::new_with_hash_id_from_sqlite_catalog_only(
```
Comment (carols10cents) [inline ingester/src/buffer_tree/partition/resolver/old_filter.rs:301]: :shocked_pikachu_face:

File: ingester/src/buffer_tree/partition/resolver/old_filter.rs [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+            PartitionId::new(42),
+            Some(ARBITRARY_PARTITION_HASH_ID.clone()),
+            ARBITRARY_TABLE_ID,
+            ARBITRARY_PARTITION_KEY.clone(),
+            vec![],
+            None,
+        );
+        let want_id = p.transition_partition_id().clone();
+
+        // Initialise a filter with the target partition.
```
Comment (carols10cents) [inline ingester/src/buffer_tree/partition/resolver/old_filter.rs:394]:
```suggestion
        // Initialise a filter not containing the target partition.
```

File: ingester/src/buffer_tree/partition/resolver/old_filter.rs [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+            Arc::new(MemCatalog::new(Arc::clone(&metrics))),
+            BackoffConfig::default(),
+            Duration::MAX,
+            metrics,
+            std::iter::empty(),
+        );
+
+        // Ask the filter for the partition.
+        //
+        // If the filter is successful and determining this is a hash ID, no
```
Comment (carols10cents) [inline ingester/src/buffer_tree/partition/resolver/old_filter.rs:407]:
```suggestion
        // If the filter is successful in determining this is a hash ID, no
```

**Review by carols10cents — APPROVED** [review]
Some comment nits I can resolve, and a big caveat that the `OldPartitionBloomFilter` isn't actually used anywhere yet, so does it really *close* https://github.com/influxdata/influxdb_iox/issues/8066? I'm going to do the hooking up though, so approved.

**Pull request merged**